Access tokens

dogfish182 opened the pull request:
Hi there, 

I've added 2 methods to create and revoke access tokens. I also needed to modify _generic_http_method_request a little bit to allow always returning the full response so I could handle errors easier in my methods.

anancarv commented:
Hi @dogfish182 ,

Thanks for your contribution. 
However, as you can see, we use pre-commit hooks for for identifying code issues before submission to code review.  Therefore, for your modifications to be reviewed and merged, could you please do the following:

pip install pre-commit
pre-commit install-hooks

Then, for each commit, the following checks will be made:
-  black
-  mypy
-  detect-secrets

dogfish182 commented:
> Hi @dogfish182 ,
> 
> Thanks for your contribution.
> However, as you can see, we use pre-commit hooks for for identifying code issues before submission to code review. Therefore, for your modifications to be reviewed and merged, could you please do the following:
> 
> pip install pre-commit
> pre-commit install-hooks
> 
> Then, for each commit, the following checks will be made:
> 
> * black: Python code formatter
> * mypy: Python type checker
> * detect-secrets: To check if your code doesn't contain secrets

Hi @anancarv 

Yep saw that and was just reviewing. Linting looks ok now 👍

anancarv commented:
Great :smiley: , I saw your modifications. 
The codacy issue is normal, we use an env variable to send coverage results to codacy. But since, your PR is from a fork. You do not have access to our env variables. I'll have a look at it

anancarv commented on pyartifactory/objects.py:
```diff
@@ -1,684 +1,745 @@
+        user_name: str,
+        expires_in: int = 3600,
+        refreshable: bool = False,
+        groups: list = None,
```
It would be better to have a single argument **access_token: AccessTokenModel**

anancarv commented on pyartifactory/objects.py:
```diff
@@ -1,684 +1,745 @@
+            "username": user_name,
+            "expires_in": expires_in,
+            "refreshable": refreshable,
+        }
```
No need to have this section if you have an argument **access_token: AccessTokenModel**

anancarv commented on pyartifactory/objects.py:
```diff
@@ -1,684 +1,745 @@
+            return AccessTokenModel(**response.json())
+        raise InvalidTokenDataException(
+            response.json().get("error_description", "Unknown error")
+        )
```
```
try: 
       response = self._post(f"api/{self._uri}/token", json=access_token.json())
       return AccessTokenModel(**response.json())
except requests.exceptions.HTTPError as e:
       raise InvalidTokenDataException(str(e))
```

anancarv commented on pyartifactory/objects.py:
```diff
@@ -1,684 +1,745 @@
+        """
+        if not any([token, token_id]):
+            logging.error("Neither a token or a token id was specified")
+            raise InvalidTokenDataException
```
Could you please add a message a comment to the exception. For example, InvalidTokenDataException("Neither a token or a token id was specified")

anancarv commented on pyartifactory/objects.py:
```diff
@@ -1,684 +1,745 @@
+            logging.info("Token revoked successfully, or token did not exist")
+            return True
+        logging.error("Token revocation unsuccessful, response was %s", response.text)
+        return False
```
```
try:
        response = self._post(f"api/{self._uri}/token/revoke", data=payload)
        logging.info("Token revoked successfully, or token did not exist")
        return True
except requests.exceptions.HTTPError as e:
        logging.error("Token revocation unsuccessful)
        raise InvalidTokenDataException(str(e))
```

anancarv commented on pyartifactory/objects.py:
```diff
@@ -1,684 +1,745 @@
+        :param kwargs: Additional parameters to add the request
+        :return: An HTTP response
+        """
+        raise_for_status = kwargs.pop("raise_for_status", True)
```
I'm not sure about this feature, it would require rewriting all existing methods. I think that we should keep the current way of raising errors and if we want to modify it, another pull request should be made.

dogfish182 replied:
actually I did it this way because it defaults to true, so it will not break your interface and will allow me to bypass raising for status when it is required. 

the reason I did this is because raising_for_status here supresses the real cause of the error and leaves you guessing. when a token can't be provisioned, If I don't pass on the response its ambiguous as to what is going on. 

example 

`art.security.create_access_token(user_name='bobby')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/ghawker/code/python/python-artifactory-fork/pyartifactory/objects.py", line 227, in create_access_token
    response.json().get("error_description", "Unknown error")
pyartifactory.exception.InvalidTokenDataException: Insufficient scope: 'api:*'`

here above we can see that a transient user gets created and you must pass groups or token generation doesn't work. with this call

`response = self._post(
            f"api/{self._uri}/token/revoke", data=payload, raise_for_status=False
        )`

I'm able to collect and handle the error correctly (it's useful for the user of the library). 

If I make the same call and pass no optional raise_for_status parameter like this 

`response = self._post(
            f"api/{self._uri}/token/revoke", data=payload
        )`
then you can see that the interface isn't broken (and no method rewriting needs to take place)

`raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: 'myurl.artifactory.com'`

of course this is your code and I can edit this out and submit as seperate PR if you like, but I do think this demonstrates both backwards compatability and a value addition for better/easier error handling.

anancarv replied:
It's actually a great feature, thanks for your explanation :ok_hand: . As you said, it can be useful to the users of the library and even think that I could use that in other methods :+1:

nymous replied:
I read your explanation @dogfish182, and I find the idea and implementation quite nice, but it feels weird to take it as a `kwargs` parameter. Since we know how to handle it and are not passing it to the actual HTTP call, I think it could be changed to a default parameter of the method. That way it would be better documented and discoverable, while still being optional.

That would change it to
```diff
     def _generic_http_method_request(
-        self, method: str, route: str, **kwargs
+        self, method: str, route: str, raise_for_status: bool = True, **kwargs
     ) -> Response:
         """
         :param method: HTTP method to use
         :param route: API Route
         :param kwargs: Additional parameters to add the request
         :return: An HTTP response
         """
-        raise_for_status = kwargs.pop("raise_for_status", True)
         http_method = getattr(self.session, method)
```

anancarv commented on pyartifactory/objects.py:
```diff
@@ -1,684 +1,745 @@
+        )
+        if raise_for_status:
+            response.raise_for_status()
+        return response
```
Regarding previous comment, you can replace this section by just **response.raise_for_status()**

dogfish182 replied:
yep, see previous comment for my rationale, and let me know if you want this seperated to a new PR.

anancarv commented on pyartifactory/models/Auth.py:
```diff
@@ -16,3 +16,11 @@ class ApiKeyModel(BaseModel):
 
 class PasswordModel(BaseModel):
     password: SecretStr
+
```
It would be great to an enum for all possible scopes, I can see in the [Artifactory api](https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#ArtifactoryRESTAPI-CreateToken) that there are 3 possible scopes.
```
class ScopeEnum(str, Enum):
    member_of_groups = "member-of-groups"
    ...
```

anancarv commented on pyartifactory/models/Auth.py:
```diff
@@ -16,3 +16,11 @@ class ApiKeyModel(BaseModel):
+class AccessTokenModel(BaseModel):
+    access_token: str
+    expires_in: Optional[int] = 0
+    scope: Optional[str] = None
```
I would rather define the scope variable as follows:

`scope: Dict[ScopeEnum, str]`

anancarv commented on pyartifactory/models/Auth.py:
```diff
@@ -16,3 +16,11 @@ class ApiKeyModel(BaseModel):
+
+class AccessTokenModel(BaseModel):
+    access_token: str
+    expires_in: Optional[int] = 0
```
Could you please set this variable to 3600

dogfish182 replied:
done

anancarv replied:
:+1:

anancarv commented:
Hi @dogfish182 , besides the comments in your PR. Can you also update the README with how to create/revoke an access_token and add tests on these two features :)

anancarv commented:
Hi @dogfish182 ,
Here is a proposition of unit tests for your **create_access_token** and **revoke_token** methods. Let me know if you are ok with them :wink:

```
@responses.activate
def test_create_access_token():
    responses.add(responses.POST, f"{URL}/api/security/token", status=200, json={
        "access_token": "<the access token>",
        "expires_in": 3600,
        "scope": "api:* member-of-groups:g1, g2",
        "token_type": "Bearer",
        "refresh_token": False
    })

    artifactory_security = ArtifactorySecurity(AuthModel(url=URL, auth=AUTH))
    access_token = artifactory_security.create_access_token(user_name="my-username", expires_in=3600, refreshable=False, groups=["g1", "g2"])
    assert access_token.refresh_token is False
    assert access_token.scope == "api:* member-of-groups:g1, g2"


@responses.activate
def test_revoke_access_token_success():
    responses.add(responses.POST, f"{URL}/api/security/token/revoke", status=200)

    artifactory_security = ArtifactorySecurity(AuthModel(url=URL, auth=AUTH))
    result = artifactory_security.revoke_access_token(token="my-token")
    assert result is True


@responses.activate
def test_revoke_access_token_fail_no_token_provided():
    responses.add(responses.POST, f"{URL}/api/security/token/revoke", status=400)

    artifactory_security = ArtifactorySecurity(AuthModel(url=URL, auth=AUTH))
    with pytest.raises(InvalidTokenDataException):
        result = artifactory_security.revoke_access_token()
        assert result is False
```

anancarv commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/test_security.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ananias/python-artifactory/pullRequest?prid=4767684&bid=15891726)

dogfish182 commented:
> ```diff
> - tests/test_security.py  1
> ```

Ok i added your tests and ran them locally, also updated the readme to include details. 

you need to fix the pipeline for codacy for me I think, but this should round out all proposed changes?

anancarv commented:
> > ```diff
> > - tests/test_security.py  1
> > ```
> 
> Ok i added your tests and ran them locally, also updated the readme to include details.
> 
> you need to fix the pipeline for codacy for me I think, but this should round out all proposed changes?

I'm working on the codacy issue. It should be fixed soon. However, the linting failed on your last commit (black issue). Could you please fix it ?

```
$ pre-commit run --all-files
black................................................................................Failed
- hook id: black
- files were modified by this hook
reformatted /home/travis/build/anancarv/python-artifactory/tests/test_security.py
All done! ✨ 🍰 ✨
1 file reformatted, 16 files left unchanged.
mypy.................................................................................Passed
Detect secrets.......................................................................Passed
bandit...............................................................................Passed
Insert a table of contents in Markdown files, like a README.md.......................Passed
The command "pre-commit run --all-files" exited with 1.
```

anancarv commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/test_security.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ananias/python-artifactory/pullRequest?prid=4767684&bid=15891726)

dogfish182 commented:
> > > ```diff
> > > - tests/test_security.py  1
> > > ```
> > 
> > 
> > Ok i added your tests and ran them locally, also updated the readme to include details.
> > you need to fix the pipeline for codacy for me I think, but this should round out all proposed changes?
> 
> I'm working on the codacy issue. It should be fixed soon. However, the linting failed on your last commit (black issue). Could you please fix it ?
> 
> ```
> $ pre-commit run --all-files
> black................................................................................Failed
> - hook id: black
> - files were modified by this hook
> reformatted /home/travis/build/anancarv/python-artifactory/tests/test_security.py
> All done! ✨ 🍰 ✨
> 1 file reformatted, 16 files left unchanged.
> mypy.................................................................................Passed
> Detect secrets.......................................................................Passed
> bandit...............................................................................Passed
> Insert a table of contents in Markdown files, like a README.md.......................Passed
> The command "pre-commit run --all-files" exited with 1.
> ```

sorry about that! Was still getting used to your workflow (which I like btw!) I think this should be everything.

anancarv commented:
> > > > ```diff
> > > > - tests/test_security.py  1
> > > > ```
> > > 
> > > 
> > > Ok i added your tests and ran them locally, also updated the readme to include details.
> > > you need to fix the pipeline for codacy for me I think, but this should round out all proposed changes?
> > 
> > 
> > I'm working on the codacy issue. It should be fixed soon. However, the linting failed on your last commit (black issue). Could you please fix it ?
> > ```
> > $ pre-commit run --all-files
> > black................................................................................Failed
> > - hook id: black
> > - files were modified by this hook
> > reformatted /home/travis/build/anancarv/python-artifactory/tests/test_security.py
> > All done! ✨ 🍰 ✨
> > 1 file reformatted, 16 files left unchanged.
> > mypy.................................................................................Passed
> > Detect secrets.......................................................................Passed
> > bandit...............................................................................Passed
> > Insert a table of contents in Markdown files, like a README.md.......................Passed
> > The command "pre-commit run --all-files" exited with 1.
> > ```
> 
> sorry about that! Was still getting used to your workflow (which I like btw!) I think this should be everything.

Everything looks good for me. Let's just wait for @nymous to have a look and then, we can merge your code

nymous commented on pyartifactory/models/Auth.py:
```diff
@@ -16,3 +16,11 @@ class ApiKeyModel(BaseModel):
+    expires_in: Optional[int] = 3600
+    scope: Optional[str] = None
+    refresh_token: Optional[bool] = False
+    token_type: str
```
If I read the [documentation](https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#ArtifactoryRESTAPI-CreateToken), Artifactory always provides the `api:*` scope, so I guess `scope` wouldn't be Optional.
Also the `refresh_token` is a string if it exists, so it should be defined as `Optional[str] = None`.

dogfish182 replied:
indeed scope will get populated with api:* always at least. 
fixed the other as well.

dogfish182 replied:
as for the objects.py.... I don't know, I run a mac and edit exclusively in pycharm, so I don't really understand how I could have got into line ending weirdness at all honestly.

nymous commented on pyartifactory/objects.py:
```diff
@@ -1,684 +1,745 @@
+        user_name: str,
+        expires_in: int = 3600,
+        refreshable: bool = False,
+        groups: list = None,
```
Can this type be even more precise like `List[str]`? Also if it has a default value of `None`, it should be typed as `Optional[List[str]]`.

dogfish182 replied:
ok done. (I'm still getting used to mypy...)

nymous commented on pyartifactory/objects.py:
```diff
@@ -1,684 +1,745 @@
+        :param user_name: Name of the user to whom an access key should be granted. transient token created if not set.
+        :param expires_in: Expiry time for the token in seconds. For eternal tokens specify 0.
+        :param refreshable: If set to true token can be refreshed using the refresh token returned. defaults False.
+        :param groups: A list of groups the token has membership of. If username is specified, groups are implied.
```
How can the `user_name` not be defined if it doesn't have a default value in the parameters?

Also what does "If username is specified, groups are implied." mean? I don't see any special behavior in the code for this case. The Artifactory docs say that if you specify `member-of-groups:*` it will use the defined user group, is this what you wanted to do?

dogfish182 replied:
let me clarify here and I'll update the docs to match. 

if you create a token for a user that does exist in artifactory (and existing named user) you token will be created and a member of the groups that match the user. 

if you create a user that does NOT exist, the user will be 'transient' and created on the fly and groups are expected. If you create a transient user and you specify no groups at all, this will error with the following.

`token = art.security.create_access_token('non_existing_user')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/me/code/python/python-artifactory-fork/pyartifactory/objects.py", line 246, in create_access_token
    response.json().get("error_description", "Unknown error")
pyartifactory.exception.InvalidTokenDataException: Insufficient scope: 'api:*'`

now when you take an existing user which has existing group memberships, I can do this 

`token = art.security.create_access_token('docker-923-upload')
 >>>token.scope
'member-of-groups:docker-923-upload api:*'`

the group name that is returned is the group the actual existing user is a member of. 

I've rewritten the params to try clarify that.

nymous commented on tests/test_security.py:
```diff
@@ -69,3 +72,45 @@ def test_revoke_user_api_key():
+    artifactory_security = ArtifactorySecurity(AuthModel(url=URL, auth=AUTH))
+    with pytest.raises(InvalidTokenDataException):
+        result = artifactory_security.revoke_access_token()
+        assert result is False
```
Since the `revoke_access_token()` method raises an exception, the `assert` on the next line isn't run at all :sweat_smile: (you can check this by changing the line to `assert result is True`). Anyway, the method doesn't return in this case, so asserting on the result doesn't make sense.

dogfish182 replied:
removed the assert.

nymous commented:
@anancarv you can merge this whenever you want :wink:

anancarv commented:
> @anancarv you can merge this whenever you want 😉

@dogfish182 will just have to fix de conflicts with the master branch and I merge and create a new release of pyartifactory very soon.
